tc/core/polyhedral/scop.cc

```diff
@@ -440,6 +440,89 @@ detail::ScheduleTree* obtainOuterBand(detail::ScheduleTree* root) {
   }
   return tree;
 }
+
+// Check if there is at least one band with a coincident dimension
+// in the tree.
+bool hasAtLeastOneCoincidentLoop(detail::ScheduleTree* tree) {
+  if (auto band = tree->elemAs<ScheduleTreeElemBand>()) {
+    if (find(band->coincident_.begin(), band->coincident_.end(), true) !=
+            band->coincident_.end() &&
+        band->permutable_) {
+      return true;
+    }
+  }
+  auto n = tree->numChildren();
+  if (n == 0) {
+    return false;
+  } else if (n == 1) {
+    return hasAtLeastOneCoincidentLoop(tree->child({0}));
+  } else {
+    for (size_t i = 0; i < n; ++i) {
+      if (hasAtLeastOneCoincidentLoop(tree->child({i}))) {
+        return true;
+      }
+    }
+  }
+  return false;
+}
+
+// Return the outermost coincident band. These are the bands with at least
+// one coincident dimension, that are permutable, and that have no such
+// coincident band as ancestors. Some of these bands are created with zero
+// dimensions to ensure that the union of all children of all outermost
+// coincident bands is equal to the leafs. Also, these created bands are
+// put at the highest possible level of the tree, which also minimize the
+// number of bands inserted.
+
+std::vector<detail::ScheduleTree*> obtainOuterCoincidentBands(
```

Comment:
This functions looks significantly more complex than it should so let me ask 2 questions:
1. my understanding is you want a set of bands with the permutability + coincident property that cover all children. You insert empty bands along paths that cannot be covered. Is this feature independently well-tested? Given that I see no standalone tests for it I would think that it isn't; yet it seems quite a complex behavior so you def want to test it in isolation.
2. it seems you are implementing a quite complex behavior where you may pick the topmost bands along some paths and are left with injecting empty bands at various depths in various paths to children than cannot be covered initially. Is this complexity really worth it? (i.e. can you exhibit examples where this is needed and it clearly brings benefits?)

One of the most annoying things in polyhedral compilation is complexity debt and this function feels like it is adding a lot of debt to me.

Can we cut the complexity by taking the topmost band with the property you want and declare that this is the depth at which we insert zero bands along all paths?

More importantly, can we kill recursion with fire, I am 99% certain this function is generally incorrect even though I do not fully understand it. Even if you go for (what I understand is) the general behavior you want, it seems to me you should be able to:
 1. `collectBFS` and use an ordered map to remember covering nodes. For each new node if one of its ancestors is covering then you're done; othwerwise if it has the property mark it; otherwise skip;
2. `collectBFS` again and filter for nodes such that no ancestor and no descendant are marked. These are the places where you want to insert 0-bands; mark them as you find them.
3. for all nodes marked in 2, insert above. 

Note that what I propose above assumes you really want to implement "all children are covered" which I don't think `hasAtLeastOneCoincidentLoop` allows you to achieve.

Reply:
In any case, I think a lot of documentation with various tree shapes that explain what you want to return in various cases is required and the tests too.
Still I would ask again if we should just err on the side of the simplest pattern that helps (e.g. get topmost band and inject 0-bands at only that depth in the tree, if needed) achieve something.

Of course this is all contingent on the end to end PR actually helping in concrete cases which from a recent discussion with @ftynse didn't seem very clear to me.